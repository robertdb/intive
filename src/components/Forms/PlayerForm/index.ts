export * from './PlayerForm';
export * from './validation';

export interface MyFormValues {
    name: string;
    position: string;
    age: number | undefined
}

// The type of props MyForm receives
export interface MyFormProps {
    initialName?: string;
    initialPosition?: string;
    initialAge?: number | undefined,
    positions?: Array<Option>
}

interface Option {
    value: string,
    content: string
}

export interface OtherProps {
    positions?: Array<Option>
}
