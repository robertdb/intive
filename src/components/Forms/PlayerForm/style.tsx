import styled from 'styled-components';

// Forms, inputs, buttons

export const Form = styled.form`
  max-width: 700px;
  margin: 0 auto;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-around;
`;
Form.displayName = "Form";

export const Input = styled.input`
  width: 200px;
  height: 35px;
  padding-left: 20px;
  border: 1px solid #ccc;
  background-color: #fff;
`;
Input.displayName = "Input";

export const Select = styled.select`
`;
Select.displayName = 'SelectPosition';

export const Button = styled.button`
  width: 100px;
  height: 35px;
  background-color: #5995ef;
  color: #fff;
  border-radius: 3px;
`;
Button.displayName = 'ButtonSubmit';