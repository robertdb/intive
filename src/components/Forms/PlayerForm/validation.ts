import * as Yup from 'yup';

export const SearchablePlayerSchema = Yup.object().shape({
    name: Yup.string()
        .min(2, 'Must be longer than 2 characters')
        .max(20, 'Nice try, nobody has a first name that long')
        .required('Required'),
    position: Yup.string()
        .notOneOf(['position'], 'Please select the position')
        .required('Required'),
    age: Yup.number()
        .min(18, 'Must be older than 18 years')
        .max(40, 'Must be younger than 40 years')
        .required('Required'),
});