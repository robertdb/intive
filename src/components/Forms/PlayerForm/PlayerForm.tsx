import * as React from 'react';
import { withFormik, FormikProps, FormikActions } from 'formik';
import { SearchablePlayerSchema } from './validation';
import { MyFormValues, MyFormProps, OtherProps } from './';
import {
    Form,
    Input,
    Select,
    Button,
} from './style'

// Aside: You may see InjectedFormikProps<OtherProps, FormValues> instead of what comes below in older code.. 
// InjectedFormikProps was artifact of when Formik only exported a HoC. 
// It is also less flexible as it MUST wrap all props (it passes them through).
export const InnerForm = (props: OtherProps & FormikProps<MyFormValues>) => {
    const {
        handleSubmit,
        isSubmitting,
        handleChange,
        handleBlur,
        values,
        errors,
        touched,
        positions
    } = props;
    let optionsPosition: Array<any> = [];
    if (positions) {
        optionsPosition = positions.map((item, index) => {
            const { value, content } = item;
            return (
                <option key={index} value={value}>
                    {content}
                </option>
            );
        });
    }
    return (
        <Form onSubmit={handleSubmit} >
            <Input
                type="text"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.name}
                name="name"
                placeholder="Player Name"
            />
            {touched.name &&
                errors.name &&
                errors.name}
            <Select
                name="position"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.position}
            >
                <option value="position">Position</option>
                {optionsPosition}
            </Select>
            {touched.position &&
                errors.position &&
                errors.position}
            <Input
                type="number"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.age}
                name="age"
            />
            {touched.age &&
                errors.age &&
                errors.age}
            {isSubmitting && <div id="submitting">Submitting</div>}
            <Button
                className={'fafa'}
                type="submit"
                disabled={isSubmitting}>
                Submit
            </Button>
        </Form>
    );
};



// Wrap our form with the using withFormik HoC
export const PlayerForm = withFormik<MyFormProps, MyFormValues>({
    // Transform outer props into form values
    mapPropsToValues: props => {
        return {
            name: props.initialName || '',
            position: props.initialPosition || 'position',
            age: props.initialAge || ''
        };
    },

    // Add a custom validation function (this can be async too!)
    validationSchema: SearchablePlayerSchema,

    handleSubmit: (values: MyFormValues, actions: FormikActions<MyFormValues>) => {
        console.log({ values, actions });
        alert(JSON.stringify(values, null, 2));
        actions.setSubmitting(false)
    }
})(InnerForm);
