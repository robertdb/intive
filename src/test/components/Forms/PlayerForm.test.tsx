import * as React from 'react';
import * as enzyme from 'enzyme';
import '../../setupTest';
import { PlayerForm, InnerForm, MyFormValues } from '../../../components/Forms/PlayerForm';

const { mount } = enzyme;

describe('Input should be update properly', () => {
  test('should name works', () => {
    const name = 'name', value = 'yuri'
    let wrapper = mount(<PlayerForm setFilterPlayers={(filters: MyFormValues )=> null}/>);
    wrapper
      .find(`input[name="${name}"]`)
      .simulate('change', { target: { name, value } })
    wrapper.update();
    expect(wrapper.find(InnerForm).props().values[name]).toBe('yuri')
  })
  test('should age works', () => {
    const age = 'age', value = 18;
    let wrapper = mount(<PlayerForm setFilterPlayers={(filters: MyFormValues )=> null}/>);
    wrapper
      .find(`input[name="${age}"]`)
      .simulate('change', { target: { name, value } })
    wrapper.update();
    expect(wrapper.find(InnerForm).props().values[name]).toBe(value)
  })
})

describe('My form default should work fine', () => {
  let wrapper: enzyme.ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;
  beforeEach(() => {
    wrapper = mount(<PlayerForm setFilterPlayers={(filters: MyFormValues )=> null}/>);
  });

  test('should position default value exists', () => {
    const name = 'position', value = 'position';
    wrapper
      .find('SelectPosition')
      .simulate('change', { target: { name, value } })
    expect(wrapper.find(InnerForm).props().values[name]).toBe(value)
  })
})

describe('My form should be update properly', () => {
  let wrapper: enzyme.ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;
  beforeEach(() => {
    const positions = [
      {value: "striker", content: "Striker"}, 
      {value: "goalkepper", content: "Goalkepper"}
    ];
    wrapper = mount(<PlayerForm positions={positions} setFilterPlayers={(filters: MyFormValues )=> null}/>);
  });

  test('should position update to striker properly', () => {
    const name = 'position', value = 'striker';
    wrapper
      .find('SelectPosition')
      .simulate('change', { target: { name, value } })
    expect(wrapper.find(InnerForm).props().values[name]).toBe(value)
  })
})

describe('Submit should work properly', () => {

  test('submits the form', () => {
    let wrapper = mount(<PlayerForm setFilterPlayers={(filters: MyFormValues )=> null}/>);
    expect(wrapper.find(InnerForm).find('#submitting')).toHaveLength(0);

    // simulate submit event. this is always sync! async calls to setState are swallowed. 
    // be careful of false positives
    wrapper.find(InnerForm).find('Form').simulate('submit', {
      preventDefault: () => { } // no op 
    });
    // Because the simulated event is 100% sync, we can use it to test the synchronous changes
    // here. Any async stuff you do inside handleSubmit will be swallowed. Thus our UI
    // will see the following changes:
    // - isSubmitting -> true (even if you set it to false asynchronously in your handleSubmit)
    // - touched: all fields

    expect(wrapper.find(InnerForm).find('#submitting')).toHaveLength(1);
    expect(wrapper.find(InnerForm).find('ButtonSubmit').props().disabled).toBe(true)
    wrapper.unmount();
  })

})