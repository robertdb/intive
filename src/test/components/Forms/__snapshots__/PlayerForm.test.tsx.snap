// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent should render correctly in "debug" mode 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithFormik(Function)
    setFilterPlayers={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "enableReinitialize": false,
      "handleSubmit": [Function],
      "initialStatus": undefined,
      "initialValues": Object {
        "age": "",
        "name": "",
        "position": "position",
      },
      "isInitialValid": false,
      "onSubmit": [Function],
      "render": [Function],
      "setFilterPlayers": [Function],
      "validate": undefined,
      "validateOnBlur": true,
      "validateOnChange": true,
      "validationSchema": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "enableReinitialize": false,
        "handleSubmit": [Function],
        "initialStatus": undefined,
        "initialValues": Object {
          "age": "",
          "name": "",
          "position": "position",
        },
        "isInitialValid": false,
        "onSubmit": [Function],
        "render": [Function],
        "setFilterPlayers": [Function],
        "validate": undefined,
        "validateOnBlur": true,
        "validateOnChange": true,
        "validationSchema": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
